src/wfs/src/main/java/org/geoserver/wfs/xml/FeatureTypeSchemaBuilder.java

```diff
@@ -192,13 +194,27 @@ public final XSDSchema buildSchemaInternal(FeatureTypeInfo[] featureTypeInfos, S
             String targetNamespace = catalog.getNamespaceByPrefix(targetPrefix).getURI();
             schema.setTargetNamespace(targetNamespace);
             schema.getQNamePrefixToNamespaceMap().put(targetPrefix, targetNamespace);
-            // add secondary namespaces from catalog
-            for (NamespaceInfo nameSpaceinfo : catalog.getNamespaces()) {
-                if (!schema.getQNamePrefixToNamespaceMap().containsKey(nameSpaceinfo.getPrefix())) {
-                    schema.getQNamePrefixToNamespaceMap().put(nameSpaceinfo.getPrefix(),
-                            nameSpaceinfo.getURI());
+
+            boolean simple = true;
+            for (int i = 0; i < featureTypeInfos.length && simple; i++) {
+                try {
+                    simple = featureTypeInfos[i].getFeatureType() instanceof SimpleFeatureType;
+                }
+                catch(IOException e) {
+                    // ignore so that broken feature types don't prevent others from continuing to work
                 }
             }
+
+            if (!simple) {
+                // add secondary namespaces from catalog
+                for (NamespaceInfo nameSpaceinfo : catalog.getNamespaces()) {
```

Comment:
Full catalog scan of namespaces.  It was here before so it's not making it worse, but this is something to look at for future improvement.